src/core/PlaneWaveBasis.jl

```diff
@@ -0,0 +1,176 @@
+    lattice = SMatrix{3, 3, T, 9}(lattice)
+    recip_lattice = 2π * inv(lattice')
+
+    @assert(mod.(grid_size, 2) == ones(3),
+            "grid_size needs to be a 3D Array with all entries odd, such that the " *
+            "symmetry of a real quantity can be properly represented.")
+    idx_DC = LinearIndices((Int.(grid_size)..., ))[ceil.(Int, grid_size ./ 2)...]
+
+    # Plan a FFT, spending some time on finding an optimal algorithm
+    # for the machine on which the computation runs
+    fft_size = nextpow.(2, grid_size)  # Optimise FFT grid
+    tmp = Array{Complex{T}}(undef, fft_size...)
+    fft_plan = plan_fft!(tmp, flags=FFTW.MEASURE)
+    ifft_plan = plan_ifft!(tmp, flags=FFTW.MEASURE)
+
+    pw = PlaneWaveBasis{T}(lattice, recip_lattice, det(lattice), det(recip_lattice),
+                           Ecut, grid_size, idx_DC, [], [], [], fft_plan, ifft_plan)
+    set_kpoints!(pw, kpoints, kweights)
+end
+
+"""
+Reset the kpoints of an existing Plane-wave basis and change the basis accordingly.
+"""
+function set_kpoints!(pw::PlaneWaveBasis{T}, kpoints, kweights; Ecut=pw.Ecut) where T
+    @assert(length(kpoints) == length(kweights),
+            "Lengths of kpoints and length of kweights need to agree")
+    @assert sum(kweights) ≈ 1 "kweights are assumed to be normalized."
+    max_E = sum(abs2, pw.recip_lattice * floor.(Int, pw.grid_size ./ 2)) / 2
+    @assert(Ecut ≤ max_E, "Ecut should be less than the maximal kinetic energy " *
+            "the grid supports (== $max_E)")
+
+    resize!(pw.kpoints, length(kpoints)) .= kpoints
+    resize!(pw.kweights, length(kweights)) .= kweights
+    resize!(pw.wfctn_basis, length(kpoints))
+
+    # Update wfctn_basis: For each k-Point select those G coords,
+    # satisfying the energy cutoff
+    for (ik, k) in enumerate(kpoints)
+        energy(q) = sum(abs2, pw.recip_lattice * q) / 2
+        p = [G for G in gcoords(pw) if energy(k + G) ≤ pw.Ecut]
+        pw.wfctn_basis[ik] = [G for G in gcoords(pw) if energy(k + G) ≤ pw.Ecut]
+    end
+    pw
+end
+
+"""
+Return a generator producing the range of wave-vector coordinates contained
+in the Fourier grid ``B_ρ`` described by the plane-wave basis.
+"""
+function gcoords(pw::PlaneWaveBasis)
+    start = -ceil.(Int, (pw.grid_size .- 1) ./ 2)
+    stop  = floor.(Int, (pw.grid_size .- 1) ./ 2)
+    ci = CartesianIndices((UnitRange.(start, stop)..., ))
+    (SVector(ci[i].I) for i in 1:length(ci))
+end
+
+
+#
+# Perform FFT
+#
+@doc raw"""
+    G_to_R!(pw::PlaneWaveBasis, f_fourier, f_real[, gcoords])
+
+Perform an in-place FFT to translate between `f_fourier`, a fourier representation
+of a function using the wave vectors specified in `gcoords` and a representation
+on the real-space density grid ``B^∗_ρ``. The function will destroy all data
+in `f_real`. If `gcoords` is absent, the full density grid ``B_ρ`` is used.
+"""
+function G_to_R!(pw::PlaneWaveBasis, f_fourier, f_real; gcoords=gcoords(pw))
```

Comment:
Add non-mutating helper function for faster prototyping? That's the downside with documentation (:-p) : if you want to be consistent you have to document the helper also, and that induces code duplication... not sure I have a good solution to that.

Reply:
Also possibly G_to_r/r_to_G would be more consistent, if we say that r is in the unit cell and R is lattice vectors. Actually I really like q=k+K and so maybe it would be better to rename G to K... let's discuss next week!

Reply:
I'll leave it unchanged until then ... it's a copy and replace anyway.